bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/Utils.java

```diff
@@ -107,4 +126,79 @@ public static String generateId(String prefix, String path) {
+        for (Resource child : resource.getChildren()) {
+            //TODO: check it's a cq:Component, used to be allowed node (filtered out by regex)
+            resourceTypes.addAll(getResourceTypes(child, request, modelFactory));
+        }
+        return resourceTypes;
+    }
+
+    public static Set<String> getXFResourceTypes(@NotNull Resource resource, @NotNull SlingHttpServletRequest request, @NotNull ModelFactory modelFactory) {
+        ExperienceFragment experienceFragment = modelFactory.getModelFromWrappedRequest(request, resource, ExperienceFragment.class);
+        if (experienceFragment != null) {
+            String fragmentPath = experienceFragment.getLocalizedFragmentVariationPath();
+            if (StringUtils.isNotEmpty(fragmentPath)) {
+                ResourceResolver resolver = resource.getResourceResolver();
+                if (resolver != null) {
+                    Resource fragmentResource = resolver.getResource(fragmentPath);
+                    if (fragmentResource != null) {
+                        return getResourceTypes(fragmentResource, request, modelFactory);
+                    }
+                }
+            }
+        }
+        return Collections.emptySet();
+    }
+
+    public static Set<String> getTemplateResourceTypes(@NotNull Page page, @NotNull SlingHttpServletRequest request, @NotNull ModelFactory modelFactory) {
+        Template template = page.getTemplate();
+        if (template != null) {
+            String templatePath = template.getPath() + AllowedComponentList.STRUCTURE_JCR_CONTENT;
+            ResourceResolver resolver = page.getContentResource().getResourceResolver();
+            if (resolver != null) {
+                Resource templateResource = resolver.getResource(templatePath);
+                if (templateResource != null) {
+                    return getResourceTypes(templateResource, request, modelFactory);
+                }
+            }
+        }
+        return Collections.emptySet();
+    }
+
+    @NotNull
+    public static Set<String> getSuperTypes(@NotNull String resourceType, @NotNull ResourceResolver resolver) {
+        Set<String> superTypes = new HashSet<>();
+        Resource resource;
+        while ((resource = resolver.getResource(resourceType)) != null) {
+            resourceType = resource.getResourceSuperType();
+            if (resourceType == null ||
+                    !superTypes.add(resourceType)) { // avoid infinite loops
+                break;
+            }
+        }
+        return superTypes;
+    }
+
+    @Nullable
+    public static ResourceResolver getComponentsResolver(ResourceResolverFactory resolverFactory) {
+        try {
+            return resolverFactory.getServiceResourceResolver(
+                    Collections.singletonMap(ResourceResolverFactory.SUBSERVICE, COMPONENTS_SERVICE));
+        } catch (LoginException e) {
+            return null;
```

Comment:
In case of LoginException we should have a log message and not just returning null

Reply:
Added logging in d9f4bb2